File: flutter_app/lib/features/transactions/domain/entities/transaction.dart
import 'package:equatable/equatable.dart';

class AppTransaction extends Equatable {
  final int id;
  final int walletId;
  final int userId;
  final int? bookingId;
  final String type;
  final double amount;
  final double balanceBefore;
  final double balanceAfter;
  final String status;
  final String description;
  final String referenceCode;
  final DateTime createdAt;

  const AppTransaction({
    required this.id,
    required this.walletId,
    required this.userId,
    this.bookingId,
    required this.type,
    required this.amount,
    required this.balanceBefore,
    required this.balanceAfter,
    required this.status,
    required this.description,
    required this.referenceCode,
    required this.createdAt,
  });

  bool get isCredit => type == 'TOP_UP' || type == 'REFUND';
  bool get isDebit => type == 'BOOKING_PAYMENT';

  @override
  List<Object?> get props => [id, type, amount, status, referenceCode];
}

class TransactionSummary extends Equatable {
  final double totalSpent;
  final double totalTopUps;
  final double totalRefunds;
  final double currentBalance;
  final int transactionCount;

  const TransactionSummary({
    required this.totalSpent,
    required this.totalTopUps,
    required this.totalRefunds,
    required this.currentBalance,
    required this.transactionCount,
  });

  @override
  List<Object?> get props =>
      [totalSpent, totalTopUps, totalRefunds, currentBalance, transactionCount];
}

class TransactionList extends Equatable {
  final List<AppTransaction> items;
  final int total;
  final int page;
  final int perPage;
  final int pages;

  const TransactionList({
    required this.items,
    required this.total,
    required this.page,
    required this.perPage,
    required this.pages,
  });

  @override
  List<Object?> get props => [items, total, page];
}

File: backend/app/schemas/transaction.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class TransactionResponse(BaseModel):
    id: int
    wallet_id: int
    user_id: int
    booking_id: Optional[int]
    type: str
    amount: float
    balance_before: float
    balance_after: float
    status: str
    description: str
    reference_code: str
    created_at: datetime

    model_config = {"from_attributes": True}


class TransactionSummary(BaseModel):
    total_spent: float
    total_top_ups: float
    total_refunds: float
    current_balance: float
    transaction_count: int


class TransactionListResponse(BaseModel):
    items: list[TransactionResponse]
    total: int
    page: int
    per_page: int
    pages: int
